### DISCUSSION TITLE
Prefix Utilization

### DISCUSSION THREAD

**Discussion by klinkeye** [netbox-community/netbox#8985] [General]
Is it expected that the parent prefix with child prefixes shows 0% utilization when the child prefixes have various values of utilization above 0%?

I would have expected the parent prefix would reflect the level of utilization of the total parent prefix with respect to the amount of child utilization.

I am a network engineer, but I might be coming at this from a different direction than others using Netbox.

**Comment by jeremystretch** [comment]
Parent prefix utilization is calculated according to the total amount of their own space consumed by allocated children. You can see examples of this on the [demo instance](https://demo.netbox.dev/ipam/prefixes/). It's impossible to say whether what you're seeing on your local instance is correct or not without more detail.

**Comment by klinkeye** [comment]
Thanks Jeremy, mine is definitely not appearing like the demo. I have a parent prefix, with child prefixes. If I mark some of those child prefixes as 100% utilized, the parent stays at 0% utilized.

Im on 3.1.8

**Reply by candlerb** [reply]
Are they in different VRFs?  If so they're considered non-overlapping.  Otherwise, can you give the exact set of steps to create the prefixes you're talking about?

**Reply by klinkeye** [reply]
> Are they in different VRFs? If so they're considered non-overlapping. Otherwise, can you give the exact set of steps to create the prefixes you're talking about?

![image](https://user-images.githubusercontent.com/83050259/161863589-7223d7ca-8cc7-42b2-8b19-1522e523dca4.png)
All in the global VRF, see screenshot

**Reply by klinkeye** [reply]
Ok so doing some trial and error. If you populate the IP addresses manually and individually within a prefix it shows the correct parent utilization, however if you select the "Mark Utilized" checkbox (treat as 100% utilized) in a child prefix's properties the parent prefix ignores that.

![image](https://user-images.githubusercontent.com/83050259/161886382-57e42c49-11e8-4f82-91b8-57623c3bc25f.png)
![image](https://user-images.githubusercontent.com/83050259/161886449-887fb8ad-aa0a-40ff-ae03-0afb9e70101e.png)

**Reply by candlerb** [reply]
Try changing the prefixes 10.200.0.0.16/, 10.200.200.0/24 and 10.200.255.0/24 from "Active" to "Container". See #1689 and #8091.

As I understand it:
* Prefixes marked as "Active" have their utilization counted based on the number of IP addresses contained, ignoring any child prefixes
* Prefixes marked as "Container" have their utilization counted based on the amount of address space carved into child prefixes - regardless of the utilization of those child prefixes.

So if you have say:

* Container Prefix 192.168.0.0/24
    * Active Prefix 192.168.0.0/26
        * Address 192.168.0.1/26
        * Address 192.168.0.2/26

then the utilization of the container is 25%, since 1/4 of its address space has been taken up by the child prefix.  Whether the actual utilization of 192.168.0.0/26 is zero or 100% or anywhere between makes no difference; the container only has 75% remaining space for allocation of further child prefixes.

The utilization of 192.168.0.0/26 will be 2/64 (or 2/62).  If you mark it as 100% utilized this shows in its own utilization, but it doesn't affect the calculation of outer prefixes. Why?

* If 192.168.0.0/24 is "Container" then its utilization is already marked at 25%, treating the whole child prefix as allocated
* If 192.168.0.0/24 is "Active" then its utilization would be 2/256 (or 2/254), based on child IP addresses only.  Since it's only looking at IP addresses, it makes no difference whether the 192.168.0.0/26 prefix is marked "100% utilized" or not.

Note: for this to work properly, a "Container" prefix should only contain prefixes, and an "Active" prefix should only contain IP addresses.

You can get into problems if a "Container" prefix contains IP addresses directly - e.g. for allocating loopbacks.  Those problems can be fixed by using "Reserved" instead of "Container": see #5925. (Anything which is not a "Container" is treated the same for the purposes of calculating utilization)

**Comment by vladsol** [comment]
It would be great if the parent prefix reflects the usage of the child prefix if it is marked as 100% utilized.

**Comment by mrkeathley** [comment]
Was there an issue generated for this? I was working on a fix internally but didn't want to replicate another solution.

**Comment by janexarturs** [comment]
is there a solution for this? @mrkeathley
